src/gui/protocolwidget.cpp

```diff
@@ -25,14 +25,115 @@
 #include "folder.h"
 #include "openfilemanager.h"
 #include "activityitemdelegate.h"
+#include "guiutility.h"
+#include "accountstate.h"
 
 #include "ui_protocolwidget.h"
 
 #include <climits>
 
 namespace OCC {
 
-bool SortedTreeWidgetItem::operator<(const QTreeWidgetItem &other) const
+QString ProtocolItem::timeString(QDateTime dt, QLocale::FormatType format)
+{
+    const QLocale loc = QLocale::system();
+    QString dtFormat = loc.dateTimeFormat(format);
+    static const QRegExp re("(HH|H|hh|h):mm(?!:s)");
+    dtFormat.replace(re, "\\1:mm:ss");
+    return loc.toString(dt, dtFormat);
+}
+
+ProtocolItem *ProtocolItem::create(const QString &folder, const SyncFileItem &item)
+{
+    auto f = FolderMan::instance()->folder(folder);
+    if (!f) {
+        return 0;
+    }
+
+    QStringList columns;
+    QDateTime timestamp = QDateTime::currentDateTime();
+    const QString timeStr = timeString(timestamp);
+    const QString longTimeStr = timeString(timestamp, QLocale::LongFormat);
+
+    columns << timeStr;
+    columns << Utility::fileNameForGuiUse(item._originalFile);
+    columns << f->shortGuiLocalPath();
+
+    // If the error string is set, it's prefered because it is a useful user message.
+    QString message = item._errorString;
+    if (message.isEmpty()) {
+        message = Progress::asResultString(item);
+    }
+    columns << message;
+
+    QIcon icon;
+    if (item._status == SyncFileItem::NormalError
+        || item._status == SyncFileItem::FatalError
+        || item._status == SyncFileItem::DetailError
+        || item._status == SyncFileItem::BlacklistedError) {
+        icon = Theme::instance()->syncStateIcon(SyncResult::Error);
+    } else if (Progress::isWarningKind(item._status)) {
+        icon = Theme::instance()->syncStateIcon(SyncResult::Problem);
+    }
+
+    if (ProgressInfo::isSizeDependent(item)) {
+        columns << Utility::octetsToString(item._size);
+    }
+
+    ProtocolItem *twitem = new ProtocolItem(columns);
+    // Warning: The data and tooltips on the columns define an implicit
+    // interface and can only be changed with care.
+    twitem->setData(0, Qt::SizeHintRole, QSize(0, ActivityItemDelegate::rowHeight()));
+    twitem->setData(0, Qt::UserRole, timestamp);
+    twitem->setIcon(0, icon);
+    twitem->setToolTip(0, longTimeStr);
+    twitem->setToolTip(1, item._file);
+    twitem->setData(2, Qt::UserRole, folder);
+    twitem->setToolTip(3, message);
+    twitem->setData(3, Qt::UserRole, item._status);
+    return twitem;
+}
+
+SyncJournalFileRecord ProtocolItem::syncJournalRecord(QTreeWidgetItem *item)
```

Comment:
Do you really want to polute the namespace for a function only used once? It's basically a oneliner since the callee also tries to get the folder.

Reply:
I think it'll be convenient in other places too and makes the code a bit more readable, in my opinon.